clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -5362,6 +5362,23 @@ LangAS CodeGenModule::GetGlobalVarAddressSpace(const VarDecl *D) {
     if (OpenMPRuntime->hasAllocateAttributeForGlobalVar(D, AS))
       return AS;
   }
+
+  if (LangOpts.HLSL) {
+    if (D == nullptr)
+      return LangAS::hlsl_private;
+
+    // Except resources (Uniform, UniformConstant) & instanglble (handles)
+    if (D->getType()->isHLSLResourceType() ||
+        D->getType()->isHLSLIntangibleType())
+      return D->getType().getAddressSpace();
+
+    if (D->getStorageClass() != SC_Static)
+      return D->getType().getAddressSpace();
+
+    LangAS AS = D->getType().getAddressSpace();
+    return AS == LangAS::Default ? LangAS::hlsl_private : AS;
+  }
+
```

Comment:
Have you considered applying the `hlsl_private` address space in the Sema (SemaHLSL::ActOnVariableDeclarator handler) onto the VarDecl QualType?

FYI, I am currently looking into adding `hlsl_constant` address space to be used for constant buffer declarations and this is probably the place where I'll add it.

Reply:
Or maybe this is the best place to do it since it is unspellable.

Reply:
I recall having tried this method first, but for some reasons moved it to Codegen, but I don't recall the full context.

I just tried moving it in sema again, and seems there is a slight issue with `this-reference-template.hlsl` since the VarDecl type is from a template, something goes wrong (but works if the AS fixup is done in codegen).